include/lock.hpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2021 Database Group, Nagoya University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CPP_UTILITY_LOCK_HPP
+#define CPP_UTILITY_LOCK_HPP
+
+#ifndef SPINLOCK_HINT
+#ifdef CPP_UTILITY_HAS_SPINLOCK_HINT
+#include <xmmintrin.h>
+#define SPINLOCK_HINT _mm_pause();  // NOLINT
+#else
+#define SPINLOCK_HINT /* do nothing */
+#endif
+#endif
+
+#include <atomic>
+
+namespace dbgroup::lock
+{
+class PessimisticLock
+{
+ public:
+  /*####################################################################################
+   * Public constructors and assignment operators
+   *##################################################################################*/
+
+  PessimisticLock(){};
+
+  /*####################################################################################
+   * Public destructors
+   *##################################################################################*/
+
+  ~PessimisticLock() = default;
+
+  /*####################################################################################
+   * Public getters
+   *##################################################################################*/
+
+  void
+  LockShared()
+  {
+    auto expected = lock_ & (~1);  // turn-off LSB
+    auto desired = expected + 2;   // increment read-counter
+    while (!lock_.compare_exchange_strong(expected, desired, std::memory_order_relaxed)) {
+      expected = expected & (~1);
+      desired = expected + 2;
+      SPINLOCK_HINT
+    }
```

Comment:
ロックを取得したり解放したりするときは競合するメモリ領域へアクセスするときなので，フェンスを細かく指定してあげる必要があると思います．ここだと，

1. 最初の`expected`を用意する時点でacquireフェンスを張って，それまでの他スレッドの書込みを見えるようにする．
2. CASが成功した場合（i.e., 共有ロックを取得できた場合）は，この処理ではロック以外の領域に副作用がないのでフェンスなしでよい．
3. CASが失敗した場合は他のスレッドによって値が書き換えられている可能性があるので，再度acquireフェンスを張って現在値を読み取る．

というふうにフェンスの設置が必要です．たぶん．

```suggestion
    auto expected = lock_.load(std::memory_order_acquire) & (~1UL);  // turn-off LSB
    auto desired = expected + 2;   // increment read-counter
    while (!lock_.compare_exchange_strong(expected, desired, std::memory_order_relaxed, std::memory_order_acquire)) {
      expected = expected & (~1UL);
      desired = expected + 2;
      SPINLOCK_HINT
    }
```

また，代入や読取りでも基本的に`load`や`store`命令を使ってフェンスを明示した方がよいです．使わない場合は全て`std::memory_order_seq_cst`が使われるので．

あと，ビット演算は扱いが色々と怖いので，64bitであることを明示（i.e., `1UL`）してます．